### PR TITLE
[Core] Newton raphson strategy - not calling InitializeSolutionStep in Predict

### LINKED ISSUE #9870: [Core] Problem when running same time step with smaller dt
**Description**
We are including what we call "roll back" in our structural solver inside the altair. The idea is simply that if SolveSolutionStep does not converge we reduce the time step in half and try again.
The problem we have is that when we reduce the time step and call InitializeSolutionStep for the next try, it is actually not doing anything as the flag mSolutionStepIsInitialized is true at that moment:

see https://github.com/KratosMultiphysics/Kratos/blob/3acacff20d1d0c815d85310c2dc58d75dab1b463/kratos/solving_strategies/strategies/residualbased_newton_raphson_strategy.h#L797

I am not sure why this flag is there, it seems its being there since the very beginning, but I think it could be removed. The only use I see of this flag is that we avoid calling InitializeSolutionStep twice when we call the Predict later. But in principle, I think we can consider that these functions are completely separated, removing the call to InitializeSolutionStep that is inside the Predict, so the flag is no longer needed.

Maybe there are some dark reasons to keep as it is right now, but in my opinion removing the flag would make it cleaner (and also would solve my problem : D )

@KratosMultiphysics/technical-committee @KratosMultiphysics/implementation-committee

### PR DESCRIPTION
**Pull request by ddiezrod**
**📝 Description**
Fixes https://github.com/KratosMultiphysics/Kratos/issues/9870 , not calling InitializeSolutionStep in Predict.
For the moment this is just to check if nothing gets broken with this change.

### CODE REVIEW

**Comment by ddiezrod** [comment]
It seems the changes do not break anything. Leaving the approval for the @KratosMultiphysics/technical-committee

**Review by loumalouomega — CHANGES_REQUESTED** [review]
I need to look at this with care, looks like a very relevant change

File: kratos/solving_strategies/strategies/residualbased_newton_raphson_strategy.h [hunk 1]
```diff
@@ -794,65 +784,61 @@ class ResidualBasedNewtonRaphsonStrategy
...
+        typename TBuilderAndSolverType::Pointer p_builder_and_solver = GetBuilderAndSolver();
+        ModelPart& r_model_part = BaseType::GetModelPart();
 
-            KRATOS_INFO_IF("ResidualBasedNewtonRaphsonStrategy", BaseType::GetEchoLevel() > 0) << "System Construction Time: "
-                << system_construction_time.ElapsedSeconds() << std::endl;
+        //set up the system, operation performed just once unless it is required
+        //to reform the dof set at each iteration
+        BuiltinTimer system_construction_time;
+        if (p_builder_and_solver->GetDofSetIsInitializedFlag() == false ||
+            mReformDofSetAtEachStep == true)
```
Comment (loumalouomega) [inline kratos/solving_strategies/strategies/residualbased_newton_raphson_strategy.h]:
```suggestion
            mReformDofSetAtEachStep)
```
Simpler

File: kratos/solving_strategies/strategies/residualbased_newton_raphson_strategy.h [hunk 2]
```diff
@@ -794,65 +784,61 @@ class ResidualBasedNewtonRaphsonStrategy
...
-            }
+        // Pointers needed in the solution
+        typename TSchemeType::Pointer p_scheme = GetScheme();
+        typename TBuilderAndSolverType::Pointer p_builder_and_solver = GetBuilderAndSolver();
+        ModelPart& r_model_part = BaseType::GetModelPart();
 
-            KRATOS_INFO_IF("ResidualBasedNewtonRaphsonStrategy", BaseType::GetEchoLevel() > 0) << "System Construction Time: "
-                << system_construction_time.ElapsedSeconds() << std::endl;
+        //set up the system, operation performed just once unless it is required
+        //to reform the dof set at each iteration
```
Comment (loumalouomega) [inline kratos/solving_strategies/strategies/residualbased_newton_raphson_strategy.h]:
```suggestion
        // To reform the dof set at each iteration
```

File: kratos/solving_strategies/strategies/residualbased_newton_raphson_strategy.h [hunk 3]
```diff
@@ -794,65 +784,61 @@ class ResidualBasedNewtonRaphsonStrategy
...
-                    << system_matrix_resize_time.ElapsedSeconds() << std::endl;
-            }
+        // Pointers needed in the solution
+        typename TSchemeType::Pointer p_scheme = GetScheme();
+        typename TBuilderAndSolverType::Pointer p_builder_and_solver = GetBuilderAndSolver();
+        ModelPart& r_model_part = BaseType::GetModelPart();
 
-            KRATOS_INFO_IF("ResidualBasedNewtonRaphsonStrategy", BaseType::GetEchoLevel() > 0) << "System Construction Time: "
-                << system_construction_time.ElapsedSeconds() << std::endl;
+        //set up the system, operation performed just once unless it is required
```
Comment (loumalouomega) [inline kratos/solving_strategies/strategies/residualbased_newton_raphson_strategy.h]:
```suggestion
        // Set up the system, operation performed just once unless it is required
```

File: kratos/solving_strategies/strategies/residualbased_newton_raphson_strategy.h [hunk 4]
```diff
@@ -794,65 +784,61 @@ class ResidualBasedNewtonRaphsonStrategy
...
+        //to reform the dof set at each iteration
+        BuiltinTimer system_construction_time;
+        if (p_builder_and_solver->GetDofSetIsInitializedFlag() == false ||
+            mReformDofSetAtEachStep == true)
+        {
+            //setting up the list of the DOFs to be solved
```
Comment (loumalouomega) [inline kratos/solving_strategies/strategies/residualbased_newton_raphson_strategy.h]:
```suggestion
            // Setting up the list of the DOFs to be solved
```

File: kratos/solving_strategies/strategies/residualbased_newton_raphson_strategy.h [hunk 5]
```diff
@@ -794,65 +784,61 @@ class ResidualBasedNewtonRaphsonStrategy
...
+            BuiltinTimer setup_dofs_time;
+            p_builder_and_solver->SetUpDofSet(p_scheme, r_model_part);
+            KRATOS_INFO_IF("ResidualBasedNewtonRaphsonStrategy", BaseType::GetEchoLevel() > 0) << "Setup Dofs Time: "
+                << setup_dofs_time.ElapsedSeconds() << std::endl;
+
+            //shaping correctly the system
```
Comment (loumalouomega) [inline kratos/solving_strategies/strategies/residualbased_newton_raphson_strategy.h]:
```suggestion
            // Shaping correctly the system
```

File: kratos/solving_strategies/strategies/residualbased_newton_raphson_strategy.h [hunk 6]
```diff
@@ -794,65 +784,61 @@ class ResidualBasedNewtonRaphsonStrategy
...
+            BuiltinTimer setup_system_time;
+            p_builder_and_solver->SetUpSystem(r_model_part);
+            KRATOS_INFO_IF("ResidualBasedNewtonRaphsonStrategy", BaseType::GetEchoLevel() > 0) << "Setup System Time: "
+                << setup_system_time.ElapsedSeconds() << std::endl;
+
+            //setting up the Vectors involved to the correct size
```
Comment (loumalouomega) [inline kratos/solving_strategies/strategies/residualbased_newton_raphson_strategy.h]:
```suggestion
            // Setting up the Vectors involved to the correct size
```

File: kratos/solving_strategies/strategies/residualbased_newton_raphson_strategy.h [hunk 7]
```diff
@@ -794,65 +784,61 @@ class ResidualBasedNewtonRaphsonStrategy
...
+            p_builder_and_solver->BuildRHS(p_scheme, r_model_part, rb);
+        }
 
-            if (mpConvergenceCriteria->GetActualizeRHSflag() == true)
-                TSparseSpace::SetToZero(rb);
+        mpConvergenceCriteria->InitializeSolutionStep(r_model_part, p_builder_and_solver->GetDofSet(), rA, rDx, rb);
 
-            mSolutionStepIsInitialized = true;
-        }
+        if (mpConvergenceCriteria->GetActualizeRHSflag() == true)
```
Comment (loumalouomega) [inline kratos/solving_strategies/strategies/residualbased_newton_raphson_strategy.h]:
```suggestion
        if (mpConvergenceCriteria->GetActualizeRHSflag())
```
Simpler

**Review by loumalouomega — COMMENTED** [review]
Looks OK, I will remove my block as soon as @KratosMultiphysics/technical-committee agrees

File: kratos/solving_strategies/strategies/residualbased_newton_raphson_strategy.h [hunk 8]
```diff
@@ -638,10 +633,6 @@ class ResidualBasedNewtonRaphsonStrategy
         if (mInitializeWasPerformed == false)
             Initialize();
```
Comment (miguelmaso) [inline kratos/solving_strategies/strategies/residualbased_newton_raphson_strategy.h:634]: To be consistent, this should be removed as well. However, I'd wait to another PR

Comment (ddiezrod) [reply kratos/solving_strategies/strategies/residualbased_newton_raphson_strategy.h:634]: good point, if we agree on this, I will continue with that later

**Comment by ddiezrod** [comment]
ping @KratosMultiphysics/technical-committee

**Review by RiccardoRossi — APPROVED** [review]
@KratosMultiphysics/technical-committee agrees about this (provided that tests pass)

**Review by loumalouomega — APPROVED** [review]
As long a test passes Okay

**Pull request merged**